File: simpleApi/src/main/java/com/lifemmo/pl/simple/api/handler/SampleApi.java
package com.lifemmo.pl.simple.api.handler;

import com.lifemmo.pl.base.protocol.Packet;
import com.lifemmo.pl.base.protocol.ReplyPacket;
import com.lifemmo.pl.base.service.*;
import com.lifemmo.pl.base.service.api.annotation.Api;
import com.lifemmo.pl.base.service.api.annotation.ApiHandler;
import com.lifemmo.pl.base.service.api.annotation.Init;
import com.lifemmo.pl.simple.Simple;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.text.SimpleDateFormat;
import java.util.Date;

@Slf4j
@Component
@Api
public class SampleApi {

    private SystemPanel systemPanel;
    private ApiBaseSender apiBaseSender;

    @Init
    public void init(SystemPanel systemPanel,ApiBaseSender apiBaseSender){
        this.systemPanel = systemPanel;
        this.apiBaseSender = apiBaseSender;
    }

    @ApiHandler(msgName = "AuthenticateReq")
    public void authenticate(String sessionInfo,Packet packet, ApiSender apiSender) throws Exception{
        Simple.AuthenticateReq req = Simple.AuthenticateReq.parseFrom(packet.buffer());
        Long accountId = req.getUserId();
        log.info("authenticate: {},{}",accountId,req.getToken());

        apiSender.authenticate(accountId,accountId.toString());


        Simple.AuthenticateRes message = Simple.AuthenticateRes.newBuilder().setUserInfo(String.valueOf(accountId)).build();
        apiSender.reply(new ReplyPacket(message));

    }

    @ApiHandler(msgName = "HelloReq")
    public void hello(String sessionInfo,Packet packet, ApiSender apiSender) throws Exception{
        Simple.HelloReq req = Simple.HelloReq.parseFrom(packet.buffer());
        log.info("hello:{},{}",req.getMessage(),apiSender.sessionInfo());
        apiSender.reply(new ReplyPacket(Simple.HelloRes.newBuilder().setMessage("hello").build()));
    }

    @ApiHandler(msgName = "SessionUpdateMsg")
    public void SessionUpdateMsg(String sessionInfo,Packet packet, ApiSender apiSender) throws Exception{
        Simple.SessionUpdateMsg msg = Simple.SessionUpdateMsg.parseFrom(packet.buffer());
        log.info("SessionUpdateMsg:{}",msg.getSessionInfo());
        apiSender.updateSession(apiSender.serviceId(),msg.getSessionInfo());

    }

    @ApiHandler(msgName = "SessionInfoReq")
    public void SessionInfoReq(String sessionInfo,Packet packet, ApiSender apiSender) throws Exception{
        log.info("SessionInfoReq:{}",sessionInfo);
        apiSender.reply(new ReplyPacket(Simple.SessionInfoRes.newBuilder().setSessionInfo(sessionInfo).build()));
    }

    @ApiHandler(msgName = "CloseSessionMsg")
    public void CloseSessionMsg(String sessionInfo,Packet packet, ApiSender apiSender) throws Exception{
        apiSender.sendToClient(new Packet(Simple.CloseSessionMsg.newBuilder().build()));

        apiSender.sessionClose(apiSender.sessionEndpoint(),apiSender.sid());

    }

    private static final SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss");

    @Scheduled(fixedRate = 5000)
    public void reportCurrentTime() {
        log.info("The time is now {}", dateFormat.format(new Date()));
    }


}

File: simpleApi/src/main/kotlin/org/ulalax/playhouse/simple/api/handler/SampleApi.kt
package org.ulalax.playhouse.simple.api.handler

import org.apache.logging.log4j.kotlin.logger
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import org.ulalax.playhouse.communicator.message.Packet
import org.ulalax.playhouse.communicator.message.ReplyPacket
import org.ulalax.playhouse.service.ApiSender
import org.ulalax.playhouse.service.Sender
import org.ulalax.playhouse.service.SystemPanel
import org.ulalax.playhouse.service.api.ApiService
import org.ulalax.playhouse.service.api.BackendHandlerRegister
import org.ulalax.playhouse.service.api.HandlerRegister
import org.ulalax.playhouse.simple.Simple.*
import org.ulalax.playhouse.simple.api.SpringContext
import java.util.*

@Component
class SampleApi : ApiService {
    private lateinit var systemPanel: SystemPanel
    private lateinit var sender: Sender
    private val log = logger()


    override suspend fun init(systemPanel: SystemPanel, sender: Sender) {
        this.systemPanel = systemPanel
        this.sender = sender
    }

    override fun instance(): ApiService {
        return SpringContext.getContext().getBean(this::class.java)
    }

    override fun handles(register: HandlerRegister,backendRegister: BackendHandlerRegister) {
        register.add(AuthenticateReq.getDescriptor().index,::authenticate)
        register.add(HelloReq.getDescriptor().index,::hello)
        register.add(CloseSessionMsg.getDescriptor().index,::closeSessionMsg)
        register.add(SendMsg.getDescriptor().index,::sendMessage)
    }

    suspend fun authenticate(packet: Packet, apiSender: ApiSender) {
        val req: AuthenticateReq = AuthenticateReq.parseFrom(packet.data())
        val accountId: Long = req.userId
        log.info("authenticate: accountId:$accountId,token:${req.token},sid:${apiSender.sid}")
        apiSender.authenticate(accountId)
        val message: AuthenticateRes = AuthenticateRes.newBuilder().setUserInfo(accountId.toString()).build()
        apiSender.reply(ReplyPacket(message))
    }

    suspend fun hello(packet: Packet, apiSender: ApiSender) {
        val req: HelloReq = HelloReq.parseFrom(packet.data())
        log.info("hello:${req.message},accountId:${apiSender.accountId},sessionEndpoint:${apiSender.sessionEndpoint},sid:${apiSender.sid}")
        apiSender.reply(ReplyPacket(HelloRes.newBuilder().setMessage("hello").build()))
    }

    suspend fun sendMessage(packet:Packet,apiSender: ApiSender){
        val recv = SendMsg.parseFrom(packet.data())
        log.info("message:${recv.message},accountId:${apiSender.accountId},sessionEndpoint:${apiSender.sessionEndpoint},sid:${apiSender.sid}")
        apiSender.sendToClient(Packet(SendMsg.getDescriptor().index,packet.movePayload()))
    }

    suspend fun closeSessionMsg( packet: Packet, apiSender: ApiSender) {
        log.info("closeSessionMsg - accountId:${apiSender.accountId},sessionEndpoint:${apiSender.sessionEndpoint},sid:${apiSender.sid}")
        apiSender.sendToClient(Packet(CloseSessionMsg.newBuilder().build()))
        apiSender.sessionClose(apiSender.sessionEndpoint, apiSender.sid)
    }

//    @Scheduled(fixedRate = 5000)
//    fun reportCurrentTime() {
//        log.info("The time is now ${dateFormat.format(Date())}")
//    }



}